cuml/bench.py

```diff
@@ -0,0 +1,528 @@
+                   verbose)
+        outer_loops = 1
+        inner_loops = 1
+        times[0] = last_warmup
+        times = times[:1]
+    else:
+        # Otherwise, actually take the timing
+        for i in range(outer_loops):
+
+            t0 = timeit.default_timer()
+            for _ in range(inner_loops):
+                val = func(*args, **kwargs)
+            t1 = timeit.default_timer()
+
+            times[i] = t1 - t0
+            total_time += times[i]
+
+            if time_limit > 0 and total_time > time_limit:
+                logverbose(f'TT={total_time:0.2f}s exceeding {time_limit}s '
+                           f'after iteration {i+1}', verbose)
+                outer_loops = i + 1
+                times = times[:outer_loops]
+                break
+
+    # We take the mean of inner loop times
+    times /= inner_loops
+    logverbose('Mean times [s]', verbose)
+    logverbose(f'{times}', verbose)
+
+    # We take the min of outer loop times
+    return np.min(times), val
+
+
+def logverbose(msg, verbose):
+    '''
+    Print msg as a verbose logging message only if verbose is True
+    '''
+    if verbose:
+        print('@', msg)
+
+
+def convert_to_numpy(data):
+    '''
+    Convert input data to numpy array
+    '''
+    if 'cudf' in str(type(data)):
+        data = data.to_pandas().values
+    elif 'pandas' in str(type(data)):
+        data = data.values
+    elif isinstance(data, np.ndarray):
+        return data
+    else:
+        raise TypeError('Unknown data format for convertion to np.ndarray')
+    return data
+
+
+def columnwise_score(y, yp, score_func):
+    y = convert_to_numpy(y)
+    yp = convert_to_numpy(yp)
+    if y.ndim + yp.ndim > 2:
+        if 1 in (y.shape + yp.shape)[1:]:
+            if y.ndim > 1:
+                y = y[:, 0]
+            if yp.ndim > 1:
+                yp = yp[:, 0]
+        else:
+            return [score_func(y[i], yp[i]) for i in range(y.shape[1])]
+    return score_func(y, yp)
+
+
+def accuracy_score(y, yp):
+    return columnwise_score(y, yp, lambda y1, y2: np.mean(y1 == y2))
+
+
+def rmse_score(y, yp):
+    return columnwise_score(
+        y, yp, lambda y1, y2: float(np.sqrt(np.mean((y1 - y2)**2))))
+
+
+def convert_data(data, dtype, data_order, data_format):
+    '''
+    Convert input data (numpy array) to needed format, type and order
+    '''
+    # Firstly, change order and type of data
+    if data_order == 'F':
+        data = np.asfortranarray(data, dtype)
+    elif data_order == 'C':
+        data = np.ascontiguousarray(data, dtype)
+
+    # Secondly, change format of data
+    if data_format == 'numpy':
+        return data
+    elif data_format == 'pandas':
+        import pandas as pd
+
+        if len(data.shape) == 1:
+            return pd.Series(data)
+        else:
+            return pd.DataFrame(data)
+    elif data_format == 'cudf':
+        import cudf
+        import pandas as pd
+
+        return cudf.DataFrame.from_pandas(pd.DataFrame(data))
```

Comment:
I think it will be correct if we first create a cupy array, and then create it to the cuda data frame.